include/dmlc/intrusive_ptr.h

```diff
@@ -0,0 +1,241 @@
+    auto copy = that;
+    std::swap(copy, *this);
+    return *this;
+  }
+  IntrusivePtr<T> &operator=(IntrusivePtr<T> &&that) {
+    std::swap(ptr_, that.ptr_);
+    return *this;
+  }
+
+  void reset() {  // NOLINT
+    Decref(ptr_);
+    ptr_ = nullptr;
+  }
+  void reset(element_type *that) { IntrusivePtr{that}.swap(*this); }  // NOLINT
+
+  element_type &operator*() const noexcept { return *ptr_; }
+  element_type *operator->() const noexcept { return ptr_; }
+  element_type *get() const noexcept { return ptr_; }  // NOLINT
+
+  explicit operator bool() const noexcept { return static_cast<bool>(ptr_); }
+
+  int32_t UseCount() noexcept {
+    return ptr_ ? IntrusivePtrRefCount(ptr_).Count() : 0;
+  }
+
+  /*
+   * \brief Helper function for swapping 2 pointers.
+   */
+  void swap(IntrusivePtr<T> &that) noexcept {  // NOLINT
+    std::swap(ptr_, that.ptr_);
+  }
+};
+
+template <class T, class U>
+bool operator==(IntrusivePtr<T> const &x, IntrusivePtr<U> const &y) noexcept {
+  return x.get() == y.get();
+}
+
+template <class T, class U>
+bool operator!=(IntrusivePtr<T> const &x, IntrusivePtr<U> const &y) noexcept {
+  return x.get() != y.get();
+}
+
+template <class T, class U>
+bool operator==(IntrusivePtr<T> const &x, U *y) noexcept {
+  return x.get() == y;
+}
+
+template <class T, class U>
+bool operator!=(IntrusivePtr<T> const &x, U *y) noexcept {
+  return x.get() != y;
+}
+
+template <class T, class U>
+bool operator==(T *x, IntrusivePtr<U> const &y) noexcept {
+  return y == x;
+}
+
+template <class T, class U>
+bool operator!=(T *x, IntrusivePtr<U> const &y) noexcept {
+  return y != x;
+}
+
+template <class T>
+bool operator<(IntrusivePtr<T> const &x, IntrusivePtr<T> const &y) noexcept {
+  return std::less<T*>{}(x.get(), y.get());
+}
+
+template <class T>
+bool operator<=(IntrusivePtr<T> const &x, IntrusivePtr<T> const &y) noexcept {
+  return std::less_equal<T*>{}(x.get(), y.get());
+}
+
+template <class T>
+bool operator>(IntrusivePtr<T> const &x, IntrusivePtr<T> const &y) noexcept {
+  return !(x <= y);
+}
+
+template <class T>
+bool operator>=(IntrusivePtr<T> const &x, IntrusivePtr<T> const &y) noexcept {
+  return !(x < y);
+}
+
+template <class E, class T, class Y>
+std::basic_ostream<E, T> &operator<<(std::basic_ostream<E, T> &os,
+                                     IntrusivePtr<Y> const &p) {
+  os << p.get();
+  return os;
+}
+
+// Constructs an object of type T and wraps it in a IntrusivePtr.
+template <typename T, typename... Args>
+IntrusivePtr<T> make_intrusive(Args&&... args) { // NOLINT
+  return IntrusivePtr<T>{new T{std::forward<Args>(args)...}};
+}
+
+/*
+ * \brief Make an intrusive pointer with Base class.  This function helps to avoid
+ *        defining reference counting function for each derived class.
+ *
+ * \tparam Base The type of base class.
+ * \tparam Sub The type of subclass.
+ * \tparam Args Type for input arguments for constructor.
+ *
+ * \param args Input arguments for constructor.
```

Comment:
Add an example code base, this function is not intuitive. Can we avoid the use of make_intrusive_base? Perhaps make every intrusive payload class to subclass a base and use https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern

Reply:
Disabling raw pointer ctor and this function will require all client code to give up on runtime dispatching, which is one of the primary reasons of using a smart pointer.  Do you really want to do that?

Reply:
If runtime dispatching is not required, users can probably just define a value instead of pointer for most of the use cases.  I think I would prefer having raw pointer ctor for intrusive ptr instead.

Reply:
I do not quite understand your comment, can you provide an example about what is the usecase? If we need to support runtime type downcasting, then that will likely mean we will need to have a type index in the intrusive ptr cell

Reply:
Runtime polymorphism.  Something like:

``` c++
class Base {
 public:
  IntrusivePtrCell mutable ref;
  friend IntrusivePtrCell &IntrusivePtrRefCount(Base const *t) noexcept {
    return t->ref;
  }

  virtual CallVirtualFunc() = 0;
};
class SubClassA : public Base {
    void CallVirtualFunc() override {
        // do something
    }
};
class SubClassB : public Base {
    void CallVirtualFunc() override {
        // do something
    }
};


intrusive_ptr<Base> ptr_a { new SubClassA };
ptr_a->CallVirtualFunc();

intrusive_ptr<Base> ptr_b { new SubClassB };
ptr_b->CallVirtualFunc();
```

The friend function and ref count are only defined in base class.

Reply:
@tqchen What do you think?

Reply:
In this case, I think we should do `make_intrusive_ptr<SubclassA>();`, and then allow automatic casting it to `intrusive_ptr<Base>`. It would be great if the API is mostly consistent with std::shared_ptr, or boost::intrusive_ptr

Reply:
I modeled the implementation after boost, which uses raw ptr for initialization.  I think for intrusive ptr the cleanest way is to just use raw ptr for initialization.  It's well established in c++ smart pointer with ownership transfer.  The `make_xxx_ptr` came later and do have there limitations.

Reply:
I still do not get the limitation of make_intrusive_ptr though, since in this case clearly we could do make_intrusive_ptr<SubclassA> and then assign to intrusive_ptr<Base> and the intention is quite clear without having to introducing a new make_intrusive_base function.

Reply:
@tqchen The `IntrusivePtrRefCount(T const *t) noexcept` function is only available for base class and not defined for subclass.  It's a friend function.

Reply:
Using above example with make_intrusive_ptr:

```
/usr/bin/ld: CMakeFiles/dmlc_unit_tests.dir/unittest_intrusive_ptr.cc.o: in function `dmlc::IntrusivePtr<dmlc::SubClassA>::Decref(dmlc::SubClassA*)':
/home/fis/Workspace/XGBoost/xgboost/dmlc-core/include/dmlc/intrusive_ptr.h:67: undefined reference to `dmlc::IntrusivePtrCell& dmlc::IntrusivePtrRefCount<dmlc::SubClassA>(dmlc::SubClassA const*)'
/usr/bin/ld: CMakeFiles/dmlc_unit_tests.dir/unittest_intrusive_ptr.cc.o: in function `dmlc::IntrusivePtr<dmlc::SubClassA>::Incref(dmlc::SubClassA*)':
/home/fis/Workspace/XGBoost/xgboost/dmlc-core/include/dmlc/intrusive_ptr.h:62: undefined reference to `dmlc::IntrusivePtrCell& dmlc::IntrusivePtrRefCount<dmlc::SubClassA>(dmlc::SubClassA const*)'
collect2: error: ld returned 1 exit status
```

Reply:
How about making all the subclasses who want to use the intrusivePtr to subclass the IntrusivePtrCell? this will allow us to avoid a type specific Cell

Reply:
That doesn't seem to be intuitive for users?  Can we accept raw pointer in constructor and `reset` method like other smart pointers?  I can add a note that intrusive ptr takes ownership. If we can accept raw pointer then `make_intrusive_base` won't be needed.

Reply:
ctor of raw ptr sounds OK, as long as we do not support operator= for raw ptr. 

I do think subclassing is not as bad when comparing to the membership approach, and we might be able to get more access control(e.g. the subclass cannot access the IncRef DecRef function).

Reply:
I vaguely remember I did something like this before.  Seems to be 2 different ways to achieve the same thing.  I don't have a preference and will stick with current implementation in next push as the code is already there.  But if you have a strong preference of using subclass instead of friend function, I can try it later.

Reply:
The main advantage of subclassing is that we could enable the assignment of intrusive_ptr<SubclassA> to intrusive_ptr<Base> (by casting to the Cell) for normal ptr storage. It also helps us to protect certain fields, e.g. keep the ref counter field private. Please consider refer to the Object class in tvm runtime

Reply:
@tqchen  Toke a brief look into the Object class in tvm,  I think the boost model of  intrusive ptr is much simpler right?  

I'm not convinced that by inheritance is better.

> It also helps us to protect certain fields, e.g. keep the ref counter field private.

I don't think any public interface class should be using intrusive ptr.